eeps/eep-0061.md

```diff
@@ -0,0 +1,255 @@
+    Status: Draft
+    Type: Standards Track
+    Created: 08-Mar-2023
+    Post-History:
+****
+EEP 61: Builtin dynamic type
+----
+
+Abstract
+========
+
+This EEP proposes a new builtin type `dynamic()` (in addition to existing
+builtin type `term()`) to make Erlang friendly to gradual type checking.
+Having a special dynamic type for gradual typing is a widely adopted industrial
+solution.
+
+Gradual typing and dynamic type
+========
+
+Erlang is one of many programming languages which started as pure dynamically
+typed languages with no static type information but then got extensions
+(or dialects) to specify type information (type hints, specs) in source code.
+
+Erlang got extended with specs through
+[EEP 8](https://www.erlang.org/eeps/eep-0008)
+(Types and function specifications).
+
+The current documentation for types and specs can be found at
+<https://www.erlang.org/doc/reference_manual/typespec.html>.
+
+Beyond being used for documentation purposes, types and specs can be used by
+tooling for static analysis. The existing tools include:
+
+- [dialyzer](https://www.erlang.org/doc/man/dialyzer.html)
+- [eqWAlizer](https://github.com/whatsapp/eqwalizer)
+
+Dialyzer uses type specs in the paradigm of success typing (documented in
+[Pull Request 6281](https://github.com/erlang/otp/pull/6281) "Document
+the meaning of specs" and in the original
+[paper](https://dl.acm.org/doi/10.1145/1292520.1292523) "A language
+for specifying type contracts in Erlang and its interaction
+with success typings"). It is worth to note that dialyzer handles specs
+very differently from traditional type checkers.
+
+eqWAlizer uses type specs to perform type checking in a traditional manner.
+
+As of now, the main industrial approach to adopt or retrofit static typing
+to dynamically typed languages is through gradual typing
+(<https://en.wikipedia.org/wiki/Gradual_typing>,
+<http://samth.github.io/gradual-typing-bib>).
+Gradual typing features a special type `dynamic` or `?`, which allows to mix
+untyped (or dynamically typed) and statically typed code in a user-friendly
+manner and also helps adopt static typing incrementally in large projects.
+Having a special dynamic type for gradual typing is a widely adopted solution.
+
+Since Erlang doesn't have a built-in dynamic type in the surface syntax,
+eqWAlizer had to introduce its own type `eqwalizer:dynamic()` with special
+semantics for the purposes of gradual typing.
+
+`term()`/`any()` and subtyping through examples
+========
+
+Before starting discussing the dynamic type, it is worth to clarify
+the meaning and semantics of the builtin Erlang types `term()` and `any()`.
+
+  - `any()` is the top type element of the subtyping lattice.
+    This is clearly stated in Erlang
+    [documentation](https://www.erlang.org/doc/reference_manual/typespec.html)
+  - `term()` is an alias to `any()`. They are the same types.
+    In the rest of the document we would use `term()` as it's more
+    Erlang-specific.
```

Comment:
There is something else to consider: `_` can also mean "I don't care about this type, infer what you want", which may be different from both `term()` and `dynamic()` in some type systems.

For Dialyzer it is all the same, but given that `_` is not that frequently used, there is potential of assigning this third use case to `_` (which is all the same in Dialyzer anyway).

Reply:
I am in strong agreement. I wanted precisely to reaffirm that `dynamic()` and `_` should be two distinct things. And I think `any()` and `term()` are used too interchangeably today to give them distinct meanings.

Overall, a type system in Erlang may have those three use cases:

* `term()`/`any()` - top type
* `dynamic()` - dynamic type
* `_` - don't make me write this part of the type, it is obvious to infer

But Dialyzer will likely treat them all the same (or _almost_ the same? good find :+1:).

Reply:
> But Dialyzer will likely treat them all the same (or almost the same? good find 👍).

Note that the find is about `erlc`, not Dialyzer.

Reply:
AFAIK, Dialyzer will consider `any() = term() = dynamic()` because the dynamic type will be simply an alias to `any()` (or `term()`), correct? If this is the case, from Dialyzer's point of view, all of them are the same.

Reply:
@ilya-klyuchnikov  New ticket for `erlc` with opaques and `_` (https://github.com/erlang/otp/issues/7015).

Reply:
To clarify above comment about Gradualizer, we also use `term()` as a dynamic type, so there is no distinction between any() and term() and `_`.

This is because in OTP, functions like `proplists:get_value/2` which has a spec like `-spec get_value(term(), [term()]) -> term()` it is in practice not the top type. If the return value of this function is passed to a function of a more restricted type, you would get a type error, even if you know that your in your proplist, the value is of a certain type. Thus, we believe it makes more sense to treat this as the dynamic type. We haven't found any occurrence of term() where it makes more sense to treat it as the top type.